imap-proto/src/types.rs

```diff
@@ -111,13 +111,101 @@ pub enum SectionPath {
     Part(Vec<u32>, Option<MessageSection>),
 }
 
+#[derive(Debug, Eq, PartialEq)]
+pub struct BodyParam<'a> {
+    pub key: &'a str,
+    pub val: &'a str,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct BodyDisposition<'a> {
+    pub disposition_type: &'a str,
+    pub param: Option<Vec<BodyParam<'a>>>,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum BodyExtension<'a> {
+    Num(u32),
+    Str(Option<&'a str>),
+    List(Vec<BodyExtension<'a>>),
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct BodyStructureText<'a> {
```

Comment:
There is a lot of duplication between the different `BodyStructure` types. Can you think about a way to reduce the duplication here? Not yet sure what that would look like, but I wouldn't want to merge it as is.